python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -740,3 +742,111 @@ def with_reconstruction_metadata(self, state_metadata: Mapping[str, Any]) -> Sel
+            fn (Callable[[AssetSpec], AssetSpec]): A function that accepts an AssetSpec and returns
+                a new AssetSpec.
+
+        Returns:
+            Definitions: A new Definitions object with the transformed assets.
+        """
+        if self.assets is None:
+            return self
+
+        new_keys_by_old_key: Dict[AssetKey, AssetKey] = {}
+
+        assets_def_new_specs: List[Tuple[AssetsDefinition, Sequence[AssetSpec]]] = []
+        new_external_asset_specs: List[AssetSpec] = []
+        new_source_assets: List[SourceAsset] = []
+
+        for el in self.assets:
+            if isinstance(el, AssetsDefinition):
+                new_specs = []
+                for spec in el.specs:
+                    new_spec = fn(spec)
+                    new_specs.append(new_spec)
+                    if new_spec.key != spec.key:
+                        new_keys_by_old_key[spec.key] = new_spec.key
+                assets_def_new_specs.append((el, new_specs))
+            elif isinstance(el, AssetSpec):
+                new_spec = fn(el)
+                if new_spec.key != el.key:
+                    new_keys_by_old_key[el.key] = new_spec.key
+                new_external_asset_specs.append(new_spec)
+            elif isinstance(el, SourceAsset):
+                new_spec = fn(el.to_asset_spec())
+                if new_spec.key != el.key:
+                    new_keys_by_old_key[el.key] = new_spec.key
+                new_source_assets.append(el.with_attributes_from_spec(new_spec))
+            elif isinstance(el, CacheableAssetsDefinition):
+                raise DagsterInvalidDefinitionError(
+                    "Can't use map_asset_specs on Definitions objects that contain "
+                    "CacheableAssetsDefinitions."
+                )
+            else:
+                check.failed(f"Unexpected asset type: {type(el)}")
+
+        # build result_assets
+        result_assets = []
+
+        for assets_def, new_specs in assets_def_new_specs:
+            new_asset_specs_replaced_deps = [
+                replace_asset_spec_dep_asset_keys(spec, new_keys_by_old_key) for spec in new_specs
+            ]
+            result_assets.append(
+                assets_def.with_replaced_asset_specs(
+                    specs=new_asset_specs_replaced_deps,
+                    new_asset_keys_by_old_asset_key=new_keys_by_old_key,
+                )
+            )
+
+        for spec in new_external_asset_specs:
+            result_assets.append(replace_asset_spec_dep_asset_keys(spec, new_keys_by_old_key))
+
+        result_assets.extend(new_source_assets)
+
+        # build result_asset_checks
+        result_asset_checks = []
+        for asset_checks_def in self.asset_checks or []:
+            result_asset_checks.append(
+                asset_checks_def.with_replaced_asset_specs(
+                    specs=[],
+                    new_asset_keys_by_old_asset_key=new_keys_by_old_key,
+                )
+            )
```

Comment:
what about asset checks in assets definitions?

Reply:
ah the implementations of `with_replaced_asset_specs` handle this 👍🏻. disregard.

Reply:
Those get handled inside `AssetsDefinition.with_replaced_asset_specs`.

The `test_asset_and_check_inside_same_op` test covers this: https://github.com/dagster-io/dagster/pull/24309/files#diff-6d1c5c6122a9ca6653bf6370201fadac6997dd01d76fd7927450d27cc13cd7a1R138.